src/server/src/main/java/org/niis/xrd4j/server/serializer/AbstractServiceResponseSerializer.java

```diff
@@ -206,42 +211,60 @@ private void processRequestNode(final SOAPBodyElement body, final ServiceRespons
         }
     }
 
-    private boolean copyRequestNode(final Node node, final SOAPBodyElement body, final ServiceResponse response) {
-        for (int i = 0; i < node.getChildNodes().getLength(); i++) {
-            if (node.getChildNodes().item(i).getNodeType() == Node.ELEMENT_NODE
-                    && "request".equals(node.getChildNodes().item(i).getLocalName())) {
-                Node requestNode = (Node) node.getChildNodes().item(i).cloneNode(true);
-                Node importNode = (Node) body.getOwnerDocument().importNode(requestNode, true);
-                body.appendChild(importNode);
-                if (response.isAddNamespaceToRequest()) {
-                    LOGGER.debug("Add provider namespace to request element.");
-                    SOAPHelper.addNamespace(importNode, response);
+    private List<SOAPElement> childElementsByLocalName(SOAPElement soapElement, String localName) {
+        var childElementsIterator = soapElement.getChildElements();
+        var stream = StreamSupport.stream(Spliterators.spliteratorUnknownSize(childElementsIterator, 0), false);
+        return stream
+                .filter(node -> node.getNodeType() == Node.ELEMENT_NODE)
+                .filter(node -> node.getLocalName().equals(localName))
+                .filter(node -> node instanceof SOAPElement)
+                .map(node -> (SOAPElement) node)
+                .collect(Collectors.toList());
+    }
+
+    private boolean copyRequestNode(final SOAPElement node, final SOAPBodyElement body, final ServiceResponse response) {
+        for (Iterator<Node> it = node.getChildElements(); it.hasNext();) {
+            var childNode = it.next();
+            if (childNode.getNodeType() == Node.ELEMENT_NODE
+                    && childNode instanceof SOAPElement
+                    && "request".equals(childNode.getLocalName())) {
+                try {
+                    var childElement = (SOAPElement) childNode;
+                    childElement = body.addChildElement(childElement);
+                    if (response.isAddNamespaceToRequest()) {
+                        LOGGER.debug("Add provider namespace to request element.");
+                        SOAPHelper.addNamespace(childElement, response);
+                    }
+                    return true;
+                } catch (SOAPException e) {
+                    LOGGER.error("Failed to copy request element.", e);
+                    throw new RuntimeException(e);
                 }
-                return true;
             }
         }
         return false;
     }
 
     private void processBodyContent(final SOAPElement soapResponse, final ServiceResponse response, final SOAPEnvelope envelope) throws SOAPException {
         // Check if there's a non-technical SOAP error
+        var sr = soapResponse;
```

Comment:
I have an issue with the usage of `sr`. It's not different than previous code as well, so it might be that I'm overthinking it :)

According to the `soapelement#setElementQName` [official documentation](https://jakarta.ee/specifications/soap-attachments/3.0/apidocs/jakarta.xml.soap/jakarta/xml/soap/soapelement#setElementQName(javax.xml.namespace.QName)):
```text
 Callers should not rely on the element instance being renamed as is. Implementations could end up copying the content of the SOAPElement to a renamed instance.
```

Now, in our code-base, `SOAPHelper#addNamespace` is using the aforementioned method and the code for  `SOAPHelper#addNamespace` is actually returning the renamed `SOAPElement` instance (regardless of it's a new one or the same).

Now, with the current method's code, `AbstractServiceResponseSerializer#processBodyContent`, we update the `soapResponse` (`sr`) without returning the renamed instance.

I don't know if this is a bug we have causing issues for some people, we got lucky with the implementations returning the same `SOAPElement` instance, or I am missing something.

If possible, please verify the usage just to make sure that I am wrong 🤣

Reply:
Here, the new SOAPElement sr is actually serialised from within the same method, that's probably why it works. I checked with unit tests and I'm quite confident, that the behaviour didn't change here.